### ISSUE TITLE
XGBoost

### ISSUE THREAD

**Issue by ijyliu** [current12/Stat-222-Project#78]
XGBoost achieved the highest accuracy on the Medium Presets (medium quality) Autogluon leaderboard, and it's widely used in industry, so it would be smart to implement.

![image](https://github.com/current12/Stat-222-Project/assets/67396209/a5aa9e41-0755-4a25-9343-a94af9ad4a11)

- Copy code from the Logistic Regression folder + be sure to produce all the same outputs as logistic regression
- I moved features_columns.json to Code\Modelling so we can use it for all the models
- You will need to encode the columns as integers, be sure to save the encoder so it can be used to get back predictions
- This will need to be run on SCF, grid search is very intensive. Run it on a small sample of the data (fewer obs and features) first to debug and make sure all the outputs are as expected and formatted the same way as logistic regression.
- Ensure grid search includes "optimal" values from Autogluon listed below, except only need to include learning rate, n_estimators, max_category_levels (or something similar that limits splitting)

![image](https://github.com/current12/Stat-222-Project/assets/67396209/12221922-5b0a-4972-83e1-e0ec25f431f1)

- Add a print for the train accuracy to see if it's overfit
- The permutation tests are VERY important for feature importance, be sure to include them

**Comment by ijyliu** [comment]
suggest @current12 works on this, since most familiar with the logistic code already

**Comment by current12** [comment]
complete the first version, haven't finished the complete GridSearch

**Comment by ijyliu** [comment]
i just created a file variable_index.xlsx in https://github.com/current12/Stat-222-Project, that we want to use for the variable names across both logistic regression and xgboost. it has columns for whether variables are categorical and numeric and whether they enter as x or y in each model. there will be 12 models for rating: altman z, tabular financial + sector, tabular financial + sector + NLP for each of: logistic regression, logistic regression with previous rating, xgboost, xgboost with previous rating. you should make the xgboost ones, logistic regression ones are done now

**Comment by ijyliu** [comment]
you should just copy over the "Rating Model" folders from the logistic regression folder + the functions to XGBoost and modify those

**Comment by ijyliu** [comment]
these hyperparameters should be enough for decent performance and a fairly fast grid search:

n_estimators
learning_rate
max_depth
min_child_weight (or something similar that controls whether or not the tree is allowed to split and stops it if too few samples)

reminder to try to include the hyperparameters autogluon found as optimal to the extent possible

i also just added code to logistic regression that prints out and saves the training accuracy, so we can tell if we are overfit. this can be a real problem with XGBoost so be sure to check it carefully and make adjustments to the hyperparameter grid if needed

**Comment by current12** [comment]
the code is up-to-date.
I also add a plot_importance functions to plot the importance of the features based on the built-in function of xgboost

**Comment by ijyliu** [comment]
@current12 

it looks like you can actually use the sklearn permutation feature importance, and just pass the model

so you should try to use that so we have the same metric as the other models. just recopy the other model code again

https://mljar.com/blog/feature-importance-xgboost/, see "Permutation Based Feature Importance (with scikit-learn)"

**Comment by ijyliu** [comment]
you don't have to do the change models, i think owen can do that

for reference though, there will be 6 xgboost change models, 3 with SMOTE and no balanced class weighting, and three with balanced class weighting

please make sure your code and functions are cleaned up and in a form that's easy for him to use

**Comment by current12** [comment]
ok, and in xgboost, "scale_pos_weight" controls the balanced and imbalanced class weighting

**Comment by ijyliu** [comment]
assuming you finished the models,

copy over + modify https://github.com/current12/Stat-222-Project/tree/main/Code/Modelling/Logistic%20Regression/Figure%20and%20Table%20Creation

you need to do both files to get tables and confusion matrices
